src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/RenamerProjectTreeActionHandler.cs

```diff
@@ -0,0 +1,211 @@
+            [Import(typeof(VisualStudioWorkspace))] Workspace workspace,
+            IEnvironmentOptions environmentOptions,
+            IUserNotificationServices userNotificationServices,
+            IRoslynServices roslynServices,
+            IWaitIndicator waitService,
+            IVsOnlineServices vsOnlineServices,
+            IProjectThreadingService threadingService,
+            IVsUIService<IVsExtensibility, IVsExtensibility3> extensibility)
+        {
+            _unconfiguredProject = unconfiguredProject;
+            _projectVsServices = projectVsServices;
+            _workspace = workspace;
+            _environmentOptions = environmentOptions;
+            _userNotificationServices = userNotificationServices;
+            _roslynServices = roslynServices;
+            _waitService = waitService;
+            _vsOnlineServices = vsOnlineServices;
+            _threadingService = threadingService;
+            _extensibility = extensibility;
+        }
+
+        public void HandleRename(string oldFilePath, string newFilePath)
+        {
+            string newFileWithExtension = Path.GetFileName(newFilePath);
+            if (StringComparer.Ordinal.Compare(_oldFilePath, oldFilePath) == 0 &&
+                StringComparer.Ordinal.Compare(_newFileNameWithExtension, newFileWithExtension) == 0)
+            {
+                _semLanguegeServiceDone.Release();
+            }
+        }
+
+        public override async Task RenameAsync(IProjectTreeActionHandlerContext context, IProjectTree node, string value)
+        {
+            Requires.NotNull(context, nameof(Context));
+            Requires.NotNull(node, nameof(node));
+            Requires.NotNullOrEmpty(value, nameof(value));
+
+            // These variables are need to synchronize with Roslyn
+            _oldFilePath = node.FilePath;
+            _newFileNameWithExtension = value;
+
+            // Do not offer to rename the file in VS Online scenarios.
+            if (_vsOnlineServices.ConnectedToVSOnline)
+            {
+                return;
+            }
+
+            if (await IsAutomationFunctionAsync() || node.IsFolder)
+            {
+                // Do not display rename Prompt
+                await base.RenameAsync(context, node, value);
+                return;
+            }
+
+            // Get the list of possible actions to execute
+            string oldName = Path.GetFileNameWithoutExtension(_oldFilePath);
+            CodeAnalysis.Project? project = GetCurrentProject();
+            if (project is null)
+            {
+                return;
+            }
+
+            CodeAnalysis.Document? oldDocument = GetDocument(project, _oldFilePath);
+            if (oldDocument is null)
+            {
+                return;
+            }
+
+            var documentRenameResult = await CodeAnalysis.Rename.Renamer.RenameDocumentAsync(oldDocument, _newFileNameWithExtension, oldDocument.Folders);
```

Comment:
nit: Folders is optional, you can just remove the parameter here if you don't want them changed

Reply:
done.